data_model/derive/src/filter.rs

```diff
@@ -0,0 +1,316 @@
+            PartialOrd,
+            Ord,
+            Eq,
+            Debug,
+            Decode,
+            Encode,
+            Deserialize,
+            Serialize,
+            IntoSchema,
+            Hash,
+        )]
+        #[doc = #filter_doc]
+        #vis struct #filter_ident #generics {
+            origin_filter: #fil_opt<#orig_fil<#imp_event>>,
+            event_filter: #fil_opt<#event_filter_ident>
+        }
+
+        impl #filter_ident {
+            #[doc = #new_doc]
+            pub const fn new(
+                origin_filter: #fil_opt<#orig_fil<#imp_event>>,
+                event_filter: #fil_opt<#event_filter_ident>,
+            ) -> Self {
+                Self {
+                    origin_filter,
+                    event_filter,
+                }
+            }
+
+            /// Get `origin_filter`
+            #[inline]
+            pub const fn origin_filter(&self) -> &#fil_opt<#orig_fil<#imp_event>> {
+                &self.origin_filter
+            }
+
+            /// Get `event_filter`
+            #[inline]
+            pub const fn event_filter(&self) -> &#fil_opt<#event_filter_ident> {
+                &self.event_filter
+            }
+        }
+
+        impl Filter for #filter_ident {
+            type EventType = #imp_event;
+            fn matches(&self, event: &Self::EventType) -> bool {
+                self.origin_filter.matches(event) && self.event_filter.matches(event)
+            }
+        }
+
+        #event_filter_and_impl
+    }
+    .into()
+}
+
+/// Generates the event filter for the event. E.g. for `AccountEvent`, `AccountEventFilter`
+/// and its `impl Filter` are generated.
+fn impl_event_filter(event: &EventEnum) -> proc_macro2::TokenStream {
+    let EventEnum {
+        vis,
+        ident: event_ident,
+        generics,
+        ..
+    } = event;
+
+    let id_variants = event.generate_filter_variants_with_id_fields();
+    let event_variants = event.generate_filter_variants_with_event_fields();
+
+    let id_impls = event.generate_filter_impls_with_id_fields();
+    let event_impls = event.generate_filter_impls_with_event_fields();
+
+    let event_filter_ident = format_ident!("{}Filter", event_ident);
+    let import_path = quote! { crate::prelude };
+    let imp_event = quote! { #import_path::#event_ident };
+
+    let event_filter_doc = format!(" Event filter for {} entity", event_ident);
+
+    quote! {
+        #[derive(
+            Clone,
+            PartialEq,
+            PartialOrd,
+            Ord,
+            Eq,
+            Debug,
+            Decode,
+            Encode,
+            Deserialize,
+            Serialize,
+            IntoSchema,
+            Hash,
+        )]
+        #[allow(clippy::enum_variant_names, missing_docs)]
+        #[doc = #event_filter_doc]
+        #vis enum #event_filter_ident #generics {
+            #(#id_variants),*,
+            #(#event_variants),*
+        }
+
+        impl Filter for #event_filter_ident {
```

Comment:
What if `Filter` is not on the scope? Or another `Filter` trait is on the scope? I suggest to use fully quilified path